File: Kramer_hyperRrs.m
%Sasha Kramer
%skramer@mbari.org
%UCSB IGPMS

%%%Script to run hyperspectral reflectance model from Kramer et al. (2022)
%Runs with: betasw_ZHH2009.m, gsm_invert.m, gsm_cost.m
%Inputs: measured Rrs, chlorophyll, temp, sal, asw, A&B coeffs for aph

%Rrs is a function of IOPs, specifically absorption and backscattering. 
%Here, you will build the absorption and backscattering terms from their 
%component parts, and then match the modeled spectra to the measured 
%spectra to optimize the model fit.

%Make sure all functions are in your path / directory!

%Map the directory where you will load your data:
cd /Users/skramer/Documents/UCSB/Research/Data/HPLC_Aph_Rrs

%Load your data
load Smoothed_Rrs.mat %Rrs
load chlorophyll.mat %chl
load temp_sal.mat %T, S

%The model uses reflectance = f(IOPs) from Gordon et al. (1988) which uses
%below the surface reflectance (rrs = Lu(0-)/Ed(0-)). If you are using 
%above surface reflectance (Rrs = Lu(0+)/Ed(0+)), you will need to convert
%your reflectances before running this model, using the equation below from
%Lee et al. (2002):
rrs = Rrs/(0.52 + 1.7*Rrs);

%Define wavelengths
wave = 400:1:700;

%%First, define total absorption as a sum of seawater absorption (asw),
%phytoplankton absorption (aph) and CDOM plus other detrital matter (acdm)
%a_tot = asw + aphstar + acdm
%asw = import aw_mcf16_350_700_1nm.txt - subset for your wavelengths
asw = asw(50:end);

%aphstar = aph/chl --> aph = import A & B coefficients from
%aph_A_B_Coeffs_Sasha_RSE_paper.txt
aph = A.*chl.^B;
aphstar = aph./chl;

%acdm slope is a function of Rrs (just above surface):
acdm_s = -(0.01447 + 0.00033*Rrs490/Rrs555);
acdm = exp(-acdm_s*(wave-443));

%%Then, define backscattering as a sum of seawater backscattering (bbsw) and
%backscattering by particles (bbp)
%bb_tot = bbsw + bbp

%bsw comes from Zhang et al. (2009): 
[~,~,bsw] = betasw_ZHH2009(wave,T,[],S);
bbsw = 0.5*bsw;

%bbp slope is a function of rrs (just below surface):
bbp_s = 2.0*(1.-1.2*exp(-0.9*rrs440rrs555)); %You will need to define rrs440 and rrs555 based on your rrs data
bbp = (443./wave).^bbp_s;

%Put IOPs together
IOPs = gsm_invert(rrs,asw,bbsw,bbp,aphstar,acdm);
%outputs = chl, acdm443, bbp443

%Reconstruct Rrs
a = asw + IOPs(:,1)*aphstar + IOPs(:,2)*acdm;
bb = bbsw + IOPs(:,3)*bbp;
rrsP = bb./(a+bb);

g = [0.0949 0.0794]; %coefficients from Gordon et al. (1988)
modrrs = (g(1) + g(2)*rrsP).*rrsP;

%Residual between measured and modeled (to use for Kramer_hyperRrs.m)
rrsD = rrs-modrrs;

File: kramer_hyper_rrs.py
import numpy as np
from scipy.io import loadmat
from kramer_functions import betasw_ZHH2009, gsm_invert

def main():
    """
    This script is a Python translation of the Kramer_hyperRrs.m MATLAB script.
    It models hyperspectral remote sensing reflectance (Rrs) and calculates the
    reflectance residual between measured and modeled values.
    """
    # Load data
    print("Loading data...")
    test_data = loadmat('data/Kramer_rrs_testdata.mat')
    Rrs = test_data['Rrs']
    chl = test_data['chl']
    T = test_data['T'].flatten()
    S = test_data['S'].flatten()

    # Convert above-surface reflectance (Rrs) to below-surface (rrs)
    rrs = Rrs / (0.52 + 1.7 * Rrs)

    # Transpose for correct orientation
    rrs = rrs.T
    Rrs = Rrs.T

    wave = np.arange(400, 701)

    # Load absorption and backscattering data
    asw_data = loadmat('data/asw_all.mat')
    asw = asw_data['asw_all'][50:, 1]  # Wavelengths from 400nm onwards

    ab_coeffs = loadmat('data/AB_coeffs.mat')
    A = ab_coeffs['A'][50:, 0]
    B = ab_coeffs['B'][50:, 0]

    # Calculate acdm (absorption due to CDOM and detritus)
    Rrs490 = Rrs[90, :]
    Rrs555 = Rrs[155, :]
    acdm_s = -(0.01447 + 0.00033 * Rrs490 / Rrs555)
    acdm = np.exp(np.outer(wave - 443, acdm_s)).T

    # Calculate seawater backscattering (bbsw)
    bbsw = np.zeros((len(wave), len(T)))
    for i in range(len(T)):
        _, _, bsw_i = betasw_ZHH2009(wave, T[i], S[i], np.arange(0, 181))
        bbsw[:, i] = 0.5 * bsw_i

    # Calculate particle backscattering (bbp) slope
    rrs440 = rrs[40, :]
    rrs555 = rrs[155, :]
    bbp_s = 2.0 * (1.0 - 1.2 * np.exp(-0.9 * rrs440 / rrs555))
    bbp = (443.0 / wave[:, np.newaxis])**bbp_s

    # Invert the GSM model to get IOPs
    print("Inverting GSM model to retrieve IOPs...")
    IOPs = np.zeros((rrs.shape[1], 3))
    for i in range(rrs.shape[1]):
        IOPs[i, :] = gsm_invert(
            rrs[:, i][np.newaxis, :], asw, bbsw[:, i],
            bbp[:, i], A, B, acdm[i, :]
        )
    print("IOPs retrieved:", IOPs)

    # Reconstruct Rrs from the retrieved IOPs
    print("Reconstructing Rrs...")
    a = np.zeros_like(rrs)
    bb = np.zeros_like(rrs)
    for i in range(rrs.shape[1]):
        a[:, i] = asw + (A * IOPs[i, 0]**B) + (IOPs[i, 1] * acdm[i, :])
        bb[:, i] = bbsw[:, i] + (IOPs[i, 2] * bbp[:, i])

    rrsP = bb / (a + bb)

    g = np.array([0.0949, 0.0794])
    modrrs = (g[0] + g[1] * rrsP) * rrsP

    # Convert modeled rrs back to Rrs
    modRrs = (0.52 * modrrs) / (1 - (1.7 * modrrs))

    # Calculate the residual
    RrsD = Rrs - modRrs

    print("Rrs residual (RrsD) calculated successfully.")
    print("Shape of RrsD:", RrsD.shape)

    # To-do: Save or use the RrsD result
    # For now, just printing a sample
    print("Sample of RrsD:", RrsD[:5, :5])

if __name__ == "__main__":
    main()